internal/collector/prometheus/prometheus_collector.go

```diff
@@ -0,0 +1,898 @@
+	freshnessThresholds FreshnessThresholds
+	fetchExecutor       executor.Executor // Polling executor for background metric fetching
+	trackedVAs          sync.Map          // map[string]*TrackedVA - tracks VAs for background fetching
+	trackedModels       sync.Map          // map[string]*TrackedModel - tracks models for replica metrics background fetching
+}
+
+// TrackedVA holds information about a VA that should be fetched in background
+type TrackedVA struct {
+	ModelID         string
+	Namespace       string
+	VariantName     string
+	VA              *llmdVariantAutoscalingV1alpha1.VariantAutoscaling
+	Deployment      appsv1.Deployment
+	AcceleratorCost float64
+	LastFetch       time.Time
+}
+
+// TrackedModel holds information about a model that should have replica metrics fetched in background
+type TrackedModel struct {
+	ModelID   string
+	Namespace string
+	LastFetch time.Time
+}
+
+// getDefaultCacheConfig returns default cache configuration
+func getDefaultCacheConfig() CacheConfig {
+	return CacheConfig{
+		Enabled:             true, // enabled by default
+		TTL:                 30 * time.Second,
+		MaxSize:             0, // unlimited by default
+		CleanupInterval:     1 * time.Minute,
+		FetchInterval:       30 * time.Second, // Fetch every 30s by default
+		FreshnessThresholds: DefaultFreshnessThresholds(),
+	}
+}
+
+// NewPrometheusCollector creates a new Prometheus metrics collector with default cache config
+// Deprecated: Use NewPrometheusCollectorWithConfig instead
+func NewPrometheusCollector(promAPI promv1.API) *PrometheusCollector {
+	return NewPrometheusCollectorWithConfig(promAPI, nil)
+}
+
+// NewPrometheusCollectorWithConfig creates a new Prometheus metrics collector
+// If cacheConfig is nil, uses default cache settings
+func NewPrometheusCollectorWithConfig(promAPI promv1.API, cacheConfig *CacheConfig) *PrometheusCollector {
+	// Use provided config or defaults
+	config := getDefaultCacheConfig()
+	if cacheConfig != nil {
+		config = *cacheConfig
+	}
+
+	var metricsCache cache.MetricsCache
+
+	if config.Enabled {
+		metricsCache = cache.NewMemoryCache(config.TTL, config.MaxSize, config.CleanupInterval)
+		logger.Log.Infof("Metrics cache enabled: TTL=%v, MaxSize=%d, CleanupInterval=%v",
+			config.TTL, config.MaxSize, config.CleanupInterval)
+	} else {
+		// Use a no-op cache implementation when disabled
+		metricsCache = &cache.NoOpCache{}
+		logger.Log.Info("Metrics cache disabled")
+	}
+
+	pc := &PrometheusCollector{
+		promAPI:             promAPI,
+		k8sClient:           nil, // Will be set when available
+		cache:               metricsCache,
+		fetchInterval:       config.FetchInterval,
+		freshnessThresholds: config.FreshnessThresholds,
+	}
+
+	// Initialize background fetching executor if fetch interval is configured
+	if config.FetchInterval > 0 {
+		pc.fetchExecutor = executor.NewPollingExecutor(executor.PollingConfig{
+			Config: executor.Config{
+				OptimizeFunc: func(ctx context.Context) error {
```

Comment:
OptimizeFunc should at some point be renamed, in the PR or as a follow-up

Reply:
left a comment